ledger/participant-integration-api/src/main/scala/platform/store/dao/BufferedCommandCompletionsReader.scala

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.api.v1.command_completion_service.CompletionStreamResponse
+import com.daml.ledger.offset.Offset
+import com.daml.logging.LoggingContext
+import com.daml.metrics.Metrics
+import com.daml.platform.store.cache.EventsBuffer
+import com.daml.platform.store.dao.BufferedCommandCompletionsReader.CompletionsFilter
+import com.daml.platform.store.interfaces.TransactionLogUpdate
+import com.daml.platform.store.interfaces.TransactionLogUpdate.CompletionDetails
+import com.daml.platform.{ApplicationId, Party}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class BufferedCommandCompletionsReader(
+    bufferReader: BufferedStreamsReader[CompletionsFilter, CompletionStreamResponse]
+) extends LedgerDaoCommandCompletionsReader {
+
+  override def getCommandCompletions(
+      startExclusive: Offset,
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Source[(Offset, CompletionStreamResponse), NotUsed] =
+    bufferReader.streamUsingBuffered(
+      startExclusive,
+      endInclusive,
+      applicationId -> parties,
+      bufferSliceFilter = filterCompletions(_, parties, applicationId),
+      toApiResponse = (response: CompletionStreamResponse) => Future.successful(response),
+    )
+
+  private def filterCompletions(
+      transactionLogUpdate: TransactionLogUpdate,
+      parties: Set[Party],
+      applicationId: String,
+  ): Option[CompletionStreamResponse] = transactionLogUpdate match {
+    case TransactionLogUpdate.TransactionAccepted(_, _, _, _, _, Some(completionDetails)) =>
+      toApiCompletion(completionDetails, parties, applicationId)
+    case TransactionLogUpdate.TransactionRejected(_, completionDetails) =>
+      toApiCompletion(completionDetails, parties, applicationId)
+    case TransactionLogUpdate.TransactionAccepted(_, _, _, _, _, None) =>
+      // Completion details missing highlights submitter is not local to this participant
+      None
+  }
+
+  private def toApiCompletion(
+      completion: CompletionDetails,
+      parties: Set[Party],
+      applicationId: String,
+  ): Option[CompletionStreamResponse] = {
+    val completionHead = completion.completionStreamResponse.completions.headOption
+      .getOrElse(throw new RuntimeException("Completions must not be empty"))
+    if (
+      completionHead.applicationId == applicationId && parties.iterator
+        .exists(completion.submitters)
```

Comment:
We should iterate over whichever party list we expect to be smaller (I don't know which that would be).

Reply:
Not sure if we can make any assumption here. But my intuition, based on some examples is that we usually have one-party subscriptions so the current state is the most efficient.